File: backend/cmd/orva/adapters/python313/adapter.py
"""Orva Python adapter — universal handler loader.

Accepts a wide range of conventions so existing code from AWS Lambda,
Google Cloud Functions, Azure, FastAPI, Flask, Django, Starlette, and
generic Python deployments runs with zero changes:

    AWS Lambda      : def lambda_handler(event, context): ...
                      def handler(event, context): ...
    GCP Functions   : def main(request): ...               (Flask Request)
    Azure Functions : def main(req): ...                   (HttpRequest)
    ASGI (FastAPI,
      Starlette)    : app = FastAPI()  /  app = Starlette()
    WSGI (Flask,
      Django)       : app = Flask(__name__)  /  app = ...   (WSGI callable)
    Plain           : def handler(event): ...

Response normalisation: the adapter accepts dicts in the Orva envelope
({statusCode, headers, body}), Starlette/FastAPI Response objects, Flask
Response objects, plain strings/dicts, or any ASGI app response. Everything
ends up as a single {statusCode, headers, body} JSON payload written to
stdout for the Orva proxy.
"""

import asyncio
import importlib.util
import inspect
import json
import os
import sys
import traceback

FUNCTION_DIR = "/code"
entrypoint = os.environ.get("ORVA_ENTRYPOINT", "handler.py")
handler_path = os.path.join(FUNCTION_DIR, entrypoint)

# Preserve the real stdout for the protocol response; reroute user print().
protocol_stdout = sys.stdout
sys.stdout = sys.stderr

if FUNCTION_DIR not in sys.path:
    sys.path.insert(0, FUNCTION_DIR)

# Make the bundled `orva` SDK (kv / invoke / jobs) importable from user
# code as `from orva import kv, invoke, jobs`. /opt/orva is the dir
# adapter.py itself runs from, but Python only auto-adds it to sys.path
# when invoked as `python /opt/orva/adapter.py` AND nothing has
# rewritten sys.path[0]. Insert explicitly so the import works
# regardless of how the adapter was invoked.
if "/opt/orva" not in sys.path:
    sys.path.insert(0, "/opt/orva")


class _Context:
    """Minimal AWS-Lambda-like context object."""

    def __init__(self, event):
        hdrs = event.get("headers", {}) if isinstance(event, dict) else {}
        self.function_name = os.environ.get("ORVA_FUNCTION_NAME", "")
        self.aws_request_id = hdrs.get("x-orva-execution-id", "")
        self.invoked_function_arn = ""
        self.memory_limit_in_mb = os.environ.get("ORVA_MEMORY_MB", "")
        self.log_group_name = "orva"
        self.log_stream_name = hdrs.get("x-orva-execution-id", "")

    def get_remaining_time_in_millis(self):
        return int(os.environ.get("ORVA_TIMEOUT_MS", "30000"))


def _load_module():
    if not os.path.exists(handler_path):
        print(f"Handler not found at {handler_path}", file=sys.stderr)
        sys.exit(1)
    spec = importlib.util.spec_from_file_location("user_handler", handler_path)
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    return mod


def _resolve(mod):
    """Return (callable, style) where style is one of:
    'lambda', 'asgi', 'wsgi', 'gcp_flask_request', 'plain'.
    """
    # ASGI frameworks (FastAPI, Starlette, Quart) — `app` is an ASGI callable.
    for name in ("app", "application"):
        app = getattr(mod, name, None)
        if app is None:
            continue
        # ASGI apps have a __call__(scope, receive, send) signature and are
        # usually async. Detect by presence of `__call__` accepting 3 args.
        if callable(app):
            sig = None
            try:
                sig = inspect.signature(app.__call__ if hasattr(app, "__call__") else app)
            except (TypeError, ValueError):
                pass
            if sig and len(sig.parameters) == 3:
                return (app, "asgi")
            # Fall through — treat as WSGI (Flask, Django WSGI, etc.).
            return (app, "wsgi")

    # Function-style exports, in priority order.
    for name in ("handler", "lambda_handler", "main"):
        fn = getattr(mod, name, None)
        if callable(fn):
            return (fn, "lambda")

    return (None, None)


mod = _load_module()
handler, style = _resolve(mod)

if handler is None:
    print(
        f"Module at {handler_path} does not export a usable handler. "
        f"Expected one of: handler, lambda_handler, main, or an ASGI/WSGI `app`.",
        file=sys.stderr,
    )
    sys.exit(1)


# ── Response normalisation ─────────────────────────────────────────────

def _normalise_response(ret):
    """Convert whatever the handler returned into (status, headers, body)."""
    # Already an Orva envelope.
    if isinstance(ret, dict) and "statusCode" in ret:
        status = ret.get("statusCode", 200)
        headers = ret.get("headers", {"Content-Type": "application/json"})
        body = ret.get("body", "")
        if not isinstance(body, str):
            body = json.dumps(body)
        return (status, headers, body)

    # Starlette / FastAPI Response objects.
    if hasattr(ret, "status_code") and hasattr(ret, "body"):
        body = ret.body
        if isinstance(body, (bytes, bytearray)):
            body = body.decode("utf-8", errors="replace")
        headers = {}
        if hasattr(ret, "headers"):
            try:
                for k, v in ret.headers.items():
                    headers[k] = v
            except Exception:
                pass
        return (ret.status_code, headers, body or "")

    # Flask Response objects.
    if hasattr(ret, "status_code") and hasattr(ret, "get_data"):
        body = ret.get_data(as_text=True)
        headers = dict(ret.headers) if hasattr(ret, "headers") else {}
        return (ret.status_code, headers, body)

    # Plain string, dict, or anything else.
    if isinstance(ret, str):
        return (200, {"Content-Type": "text/plain"}, ret)
    return (200, {"Content-Type": "application/json"}, json.dumps(ret))


# ── Invocation bridges ─────────────────────────────────────────────────

def _build_flask_request(event):
    """Build a werkzeug Request (for GCP Functions / flask-style `main(request)`)."""
    try:
        from werkzeug.wrappers import Request
        from werkzeug.test import EnvironBuilder
    except ImportError:
        return None
    body = event.get("body") or ""
    builder = EnvironBuilder(
        method=event.get("method", "POST"),
        path=event.get("path", "/"),
        headers=event.get("headers", {}),
        data=body.encode("utf-8") if isinstance(body, str) else body,
    )
    return Request(builder.get_environ())


async def _call_asgi(app, event):
    """Drive an ASGI app through one request/response cycle."""
    body_bytes = (event.get("body") or "").encode("utf-8")
    headers_list = [
        (k.lower().encode(), str(v).encode())
        for k, v in (event.get("headers") or {}).items()
    ]
    path = event.get("path", "/") or "/"
    query = ""
    if "?" in path:
        path, query = path.split("?", 1)

    scope = {
        "type": "http",
        "asgi": {"version": "3.0", "spec_version": "2.3"},
        "http_version": "1.1",
        "method": event.get("method", "GET"),
        "scheme": "http",
        "path": path,
        "raw_path": path.encode(),
        "query_string": query.encode(),
        "headers": headers_list,
        "server": ("orva", 8443),
        "client": ("127.0.0.1", 0),
    }

    body_sent = False

    async def receive():
        nonlocal body_sent
        if body_sent:
            return {"type": "http.disconnect"}
        body_sent = True
        return {"type": "http.request", "body": body_bytes, "more_body": False}

    response = {"status": 200, "headers": {}, "body": b""}

    async def send(message):
        if message["type"] == "http.response.start":
            response["status"] = message["status"]
            for k, v in message.get("headers", []):
                response["headers"][k.decode()] = v.decode()
        elif message["type"] == "http.response.body":
            response["body"] += message.get("body", b"")

    await app(scope, receive, send)
    return (
        response["status"],
        response["headers"],
        response["body"].decode("utf-8", errors="replace"),
    )


def _call_wsgi(app, event):
    """Drive a WSGI app (Flask, Django WSGI) through one request cycle."""
    from io import BytesIO

    body = event.get("body") or ""
    body_bytes = body.encode("utf-8") if isinstance(body, str) else body
    path = event.get("path", "/") or "/"
    query = ""
    if "?" in path:
        path, query = path.split("?", 1)

    environ = {
        "REQUEST_METHOD": event.get("method", "GET"),
        "SCRIPT_NAME": "",
        "PATH_INFO": path,
        "QUERY_STRING": query,
        "SERVER_NAME": "orva",
        "SERVER_PORT": "8443",
        "SERVER_PROTOCOL": "HTTP/1.1",
        "wsgi.version": (1, 0),
        "wsgi.url_scheme": "http",
        "wsgi.input": BytesIO(body_bytes),
        "wsgi.errors": sys.stderr,
        "wsgi.multithread": False,
        "wsgi.multiprocess": False,
        "wsgi.run_once": True,
        "CONTENT_LENGTH": str(len(body_bytes)),
    }
    for k, v in (event.get("headers") or {}).items():
        key = "HTTP_" + k.upper().replace("-", "_")
        environ[key] = str(v)
        if k.lower() == "content-type":
            environ["CONTENT_TYPE"] = str(v)

    captured = {"status": "200 OK", "headers": []}

    def start_response(status, headers, exc_info=None):
        captured["status"] = status
        captured["headers"] = headers
        return lambda x: None

    chunks = app(environ, start_response)
    body_out = b"".join(chunks if not isinstance(chunks, (bytes, str)) else [chunks])
    if isinstance(body_out, str):
        body_out = body_out.encode("utf-8")

    status_code = int(captured["status"].split(" ", 1)[0])
    headers_dict = {k: v for k, v in captured["headers"]}
    return (status_code, headers_dict, body_out.decode("utf-8", errors="replace"))


# ── Framed stdio protocol ──────────────────────────────────────────────
# Wire format: 4-byte big-endian uint32 length, then N bytes UTF-8 JSON.
# Same on stdin (proxy → adapter) and stdout (adapter → proxy).

import struct

_stdin = sys.stdin.buffer
_stdout = protocol_stdout.buffer if hasattr(protocol_stdout, "buffer") else protocol_stdout


def _read_exact(n):
    buf = bytearray()
    while len(buf) < n:
        chunk = _stdin.read(n - len(buf))
        if not chunk:
            return None
        buf.extend(chunk)
    return bytes(buf)


def _read_frame():
    header = _read_exact(4)
    if header is None:
        return None
    (length,) = struct.unpack(">I", header)
    if length == 0:
        return {}
    payload = _read_exact(length)
    if payload is None:
        return None
    try:
        return json.loads(payload.decode("utf-8"))
    except Exception:
        return {"type": "request", "event": {"method": "POST", "path": "/", "headers": {}, "body": ""}}


def _write_frame(obj):
    body = json.dumps(obj).encode("utf-8")
    _stdout.write(struct.pack(">I", len(body)))
    _stdout.write(body)
    _stdout.flush()


def _dispatch(event):
    if style == "asgi":
        status, headers, body = asyncio.run(_call_asgi(handler, event))
        return status, headers, body
    if style == "wsgi":
        return _call_wsgi(handler, event)

    # Lambda / plain style.
    try:
        result = handler(event, _Context(event))
    except TypeError as te:
        msg = str(te)
        if "positional argument" in msg or "takes" in msg or "missing" in msg:
            try:
                result = handler(event)
            except TypeError:
                req = _build_flask_request(event)
                if req is not None:
                    result = handler(req)
                else:
                    raise
        else:
            raise

    if inspect.iscoroutine(result):
        result = asyncio.run(result)

    return _normalise_response(result)


# ── Main loop ──────────────────────────────────────────────────────────

max_reqs = int(os.environ.get("ORVA_MAX_REQUESTS", "0") or 0)
served = 0

try:
    while True:
        frame = _read_frame()
        if frame is None:
            sys.exit(0)  # stdin EOF
        ftype = frame.get("type")
        if ftype == "quit":
            _write_frame({"type": "bye"})
            sys.exit(0)
        if ftype != "request":
            continue

        event = frame.get("event") or {"method": "POST", "path": "/", "headers": {}, "body": ""}
        # Propagate call depth into the env so orva.invoke()'s SDK can
        # forward it on outbound nested calls. Without this each recursion
        # level would see depth="" and the host's depth guard never trips.
        _hdrs = (event.get("headers") or {}) if isinstance(event, dict) else {}
        _depth = _hdrs.get("x-orva-call-depth") or _hdrs.get("X-Orva-Call-Depth") or ""
        if _depth:
            os.environ["ORVA_CALL_DEPTH"] = _depth
        else:
            os.environ.pop("ORVA_CALL_DEPTH", None)
        try:
            status, headers, body = _dispatch(event)
        except Exception:
            traceback.print_exc()
            status, headers, body = (
                500,
                {"Content-Type": "application/json"},
                json.dumps({"error": "Internal function error"}),
            )
        _write_frame({"type": "response", "statusCode": status, "headers": headers, "body": body})

        served += 1
        if max_reqs > 0 and served >= max_reqs:
            _write_frame({"type": "bye"})
            sys.exit(0)
except SystemExit:
    raise
except Exception as exc:
    try:
        _write_frame({"type": "error", "fatal": True, "message": f"{type(exc).__name__}: {exc}"})
    except Exception:
        pass
    sys.exit(1)

File: backend/runtimes/node22/adapter.js
// Orva Node.js adapter — universal handler loader.
//
// Accepts a wide range of export conventions so existing code from AWS
// Lambda, Cloudflare Workers, Vercel, Next.js, Netlify, and generic
// Node/Express style deploys runs with zero changes:
//
//   AWS Lambda         : exports.handler = async (event, context) => ...
//                        exports.lambda_handler = async (event, context) => ...
//   Cloudflare Worker  : export default { fetch(request, env, ctx) { ... } }
//                        addEventListener('fetch', e => e.respondWith(...))
//   Vercel / Next API  : export default async function handler(req, res) { ... }
//   Netlify / generic  : exports.handler = async (event) => ...
//   Plain function     : module.exports = async (event) => ...
//
// The adapter normalises all of them to a { statusCode, headers, body }
// response envelope, the native Orva protocol.

const path = require('path');
const Module = require('module');

const FUNCTION_DIR = '/code';
const entrypoint = process.env.ORVA_ENTRYPOINT || 'handler.js';
const handlerPath = path.join(FUNCTION_DIR, entrypoint);

// Make the bundled `orva` SDK module (kv / invoke / jobs) resolvable
// from user code via `require('orva')`. The package lives at
// /opt/orva/node_modules/orva/; injecting that dir at the front of
// Module._nodeModulePaths ensures user modules can find it without
// the user having to install it. User-installed deps still resolve
// normally via /code/node_modules.
const _origNodeModulePaths = Module._nodeModulePaths;
Module._nodeModulePaths = function (from) {
  return ['/opt/orva/node_modules', ...(_origNodeModulePaths.call(this, from) || [])];
};

// Preserve stdout for the protocol response; reroute user output to stderr.
const originalStdoutWrite = process.stdout.write.bind(process.stdout);
const writeProtocol = (s) => originalStdoutWrite(s);
process.stdout.write = process.stderr.write.bind(process.stderr);
console.log = (...a) => process.stderr.write(a.map(String).join(' ') + '\n');
console.info = console.log;
console.debug = console.log;
console.warn = console.log;
console.error = console.log;

let mod;
try {
  mod = require(handlerPath);
} catch (err) {
  process.stderr.write(`Failed to load handler from ${handlerPath}: ${err.message}\n`);
  process.exit(1);
}

// Unwrap ESM default export shim (Babel/TS sometimes emit { default: fn }).
if (mod && typeof mod === 'object' && mod.__esModule && mod.default) {
  mod = mod.default;
}

// Resolve a callable from whatever shape the user exported.
let handler = null;
let style = null; // "lambda" | "worker" | "vercel" | "plain"

if (typeof mod === 'function') {
  handler = mod;
  // A plain function could be Lambda-style or Vercel-style — decide by arity.
  style = mod.length >= 2 ? 'vercel-or-lambda' : 'plain';
} else if (mod && typeof mod === 'object') {
  // Cloudflare Worker: { fetch(request, env, ctx) }
  if (typeof mod.fetch === 'function') {
    handler = mod.fetch;
    style = 'worker';
  } else if (typeof mod.handler === 'function') {
    handler = mod.handler;
    style = 'lambda';
  } else if (typeof mod.lambda_handler === 'function') {
    handler = mod.lambda_handler;
    style = 'lambda';
  } else if (typeof mod.main === 'function') {
    handler = mod.main;
    style = 'lambda';
  } else if (typeof mod.default === 'function') {
    handler = mod.default;
    style = 'plain';
  }
}

if (!handler) {
  process.stderr.write(
    `Module at ${handlerPath} does not export a usable handler. ` +
    `Expected one of: handler, lambda_handler, main, fetch, default, ` +
    `or a default function export.\n`
  );
  process.exit(1);
}

// ── Helpers to bridge calling conventions ──────────────────────────────

function buildLambdaContext(event) {
  const hdrs = (event && event.headers) || {};
  return {
    functionName: process.env.ORVA_FUNCTION_NAME || '',
    awsRequestId: hdrs['x-orva-execution-id'] || '',
    invokedFunctionArn: '',
    memoryLimitInMB: process.env.ORVA_MEMORY_MB || '',
    logGroupName: 'orva',
    logStreamName: hdrs['x-orva-execution-id'] || '',
    getRemainingTimeInMillis: () => Number(process.env.ORVA_TIMEOUT_MS || 30000),
  };
}

// Minimal Request/Response polyfills for Cloudflare Worker style. If the
// runtime provides them natively (Node 18+ has global fetch), prefer those.
function buildWorkerRequest(event) {
  const url = `http://localhost${event.path || '/'}`;
  if (typeof Request === 'function') {
    try {
      return new Request(url, {
        method: event.method || 'GET',
        headers: event.headers || {},
        body: ['GET', 'HEAD'].includes((event.method || 'GET').toUpperCase())
          ? undefined
          : (event.body || ''),
      });
    } catch { /* fall through to shim */ }
  }
  return {
    method: event.method || 'GET',
    url,
    headers: new Map(Object.entries(event.headers || {})),
    body: event.body || '',
    async text() { return this.body; },
    async json() { return JSON.parse(this.body || '{}'); },
    async arrayBuffer() { return Buffer.from(this.body || '').buffer; },
  };
}

async function normaliseResponse(ret) {
  // Already in Orva envelope.
  if (ret && typeof ret === 'object' && 'statusCode' in ret) {
    return {
      statusCode: ret.statusCode || 200,
      headers: ret.headers || { 'Content-Type': 'application/json' },
      body: typeof ret.body === 'string' ? ret.body : JSON.stringify(ret.body || {}),
    };
  }
  // Fetch API Response.
  if (ret && typeof ret === 'object' && typeof ret.status === 'number' && typeof ret.text === 'function') {
    const body = await ret.text();
    const headers = {};
    if (ret.headers && typeof ret.headers.forEach === 'function') {
      ret.headers.forEach((v, k) => { headers[k] = v; });
    }
    return { statusCode: ret.status, headers, body };
  }
  // Anything else → JSON-encode as the body.
  return {
    statusCode: 200,
    headers: { 'Content-Type': 'application/json' },
    body: typeof ret === 'string' ? ret : JSON.stringify(ret ?? null),
  };
}

// Vercel/Next-style (req, res) detection: wrap to capture the response.
function invokeVercelStyle(fn, event) {
  return new Promise((resolve, reject) => {
    const req = {
      method: event.method || 'GET',
      url: event.path || '/',
      headers: event.headers || {},
      body: (() => {
        try { return JSON.parse(event.body || 'null'); } catch { return event.body; }
      })(),
      query: {},
    };
    let statusCode = 200;
    const headers = {};
    let body = '';
    const res = {
      status(c) { statusCode = c; return this; },
      setHeader(k, v) { headers[k] = v; return this; },
      getHeader(k) { return headers[k]; },
      json(o) { headers['Content-Type'] = 'application/json'; body = JSON.stringify(o); this.end(); },
      send(x) { body = typeof x === 'string' ? x : JSON.stringify(x); this.end(); },
      write(x) { body += typeof x === 'string' ? x : String(x); },
      end(x) {
        if (x !== undefined) body = typeof x === 'string' ? x : String(x);
        resolve({ statusCode, headers, body });
      },
    };
    Promise.resolve()
      .then(() => fn(req, res))
      .catch(reject);
  });
}

// ── Framed stdio protocol ──────────────────────────────────────────────
// Wire format: 4-byte big-endian uint32 length, then N bytes of UTF-8 JSON.
// Applied symmetrically to stdin (proxy → adapter) and stdout (adapter →
// proxy). Length-prefix chosen over JSONL because it is binary-safe and
// handles the full 6 MB MaxBodyBytes without escape gymnastics.

async function dispatch(event) {
  if (style === 'worker') {
    const req = buildWorkerRequest(event);
    const env = { ...process.env };
    const ctx = { waitUntil: () => {}, passThroughOnException: () => {} };
    return await handler(req, env, ctx);
  }
  if (style === 'vercel-or-lambda') {
    // Two-arg function: try Lambda first (returns a value). If it returns
    // undefined OR throws a TypeError trying to treat ctx as `res`, assume
    // Vercel (req, res) style and replay via invokeVercelStyle.
    const ctx = buildLambdaContext(event);
    let lambdaResult;
    let lambdaErr;
    try {
      lambdaResult = await handler(event, ctx);
    } catch (e) {
      lambdaErr = e;
    }
    const looksLikeVercelMiss =
      lambdaErr &&
      (lambdaErr instanceof TypeError) &&
      /is not a function|Cannot read propert/i.test(String(lambdaErr.message));
    if (lambdaResult === undefined || looksLikeVercelMiss) {
      return await invokeVercelStyle(handler, event);
    }
    if (lambdaErr) throw lambdaErr;
    return lambdaResult;
  }
  if (style === 'lambda') return await handler(event, buildLambdaContext(event));
  return await handler(event);
}

function writeFrame(obj) {
  const body = Buffer.from(JSON.stringify(obj), 'utf-8');
  const hdr = Buffer.alloc(4);
  hdr.writeUInt32BE(body.length, 0);
  // Node's process.stdout.write returns false if the kernel pipe buffer
  // is full (backpressure from the proxy). We ignore the return — the
  // next write blocks naturally until drain. EPIPE is rethrown to the
  // caller so streaming loops can stop iterating on client disconnect.
  originalStdoutWrite(hdr);
  originalStdoutWrite(body);
}

// v0.4 C1: streaming helpers — translate user-yielded values into
// `chunk` frames. data may be Buffer | string | Uint8Array | object.
function streamChunk(data) {
  let buf;
  if (data == null) {
    buf = Buffer.alloc(0);
  } else if (Buffer.isBuffer(data)) {
    buf = data;
  } else if (data instanceof Uint8Array) {
    buf = Buffer.from(data.buffer, data.byteOffset, data.byteLength);
  } else if (typeof data === 'string') {
    buf = Buffer.from(data, 'utf-8');
  } else {
    buf = Buffer.from(JSON.stringify(data), 'utf-8');
  }
  writeFrame({ type: 'chunk', data: buf.length === 0 ? '' : buf.toString('base64') });
}

function looksLikeHead(item) {
  if (!item || typeof item !== 'object') return null;
  if (!('statusCode' in item)) return null;
  return {
    status: item.statusCode || 200,
    headers: item.headers || { 'Content-Type': 'text/plain' },
    body: 'body' in item ? item.body : null,
  };
}

// streamIterable consumes any sync/async iterable and emits the
// streaming protocol exchange. When streamingEnabled is false we
// buffer everything into a single response frame for back-compat —
// operators flipping the system_config flag get pre-C1 behaviour
// without redeploying.
async function streamIterable(iterable, streamingEnabled, keepaliveMs) {
  if (!streamingEnabled) {
    let head = null;
    const parts = [];
    const collect = (item) => {
      if (head === null) {
        const detected = looksLikeHead(item);
        if (detected) {
          head = { status: detected.status, headers: detected.headers };
          if (detected.body != null) parts.push(typeof detected.body === 'string' ? detected.body : String(detected.body));
          return;
        }
        head = { status: 200, headers: { 'Content-Type': 'text/plain' } };
      }
      if (Buffer.isBuffer(item)) parts.push(item.toString('utf-8'));
      else if (item instanceof Uint8Array) parts.push(Buffer.from(item).toString('utf-8'));
      else if (typeof item === 'string') parts.push(item);
      else parts.push(String(item));
    };
    for await (const item of iterable) collect(item);
    if (head === null) head = { status: 200, headers: { 'Content-Type': 'text/plain' } };
    writeFrame({
      type: 'response', statusCode: head.status,
      headers: head.headers, body: parts.join(''),
    });
    return;
  }

  let headSent = false;
  let lastEmit = Date.now();
  let hbTimer = null;

  const sendHead = (status, headers) => {
    if (headSent) return;
    writeFrame({ type: 'response_start', statusCode: status, headers });
    headSent = true;
  };

  const startHeartbeat = () => {
    if (hbTimer) return;
    // setInterval fires regardless of how busy the loop is. The check
    // against lastEmit prevents an empty chunk from racing a real one;
    // worst case we emit one extra empty chunk per period, which is
    // harmless on the wire.
    hbTimer = setInterval(() => {
      if (Date.now() - lastEmit >= keepaliveMs) {
        try {
          writeFrame({ type: 'chunk', data: '' });
          lastEmit = Date.now();
        } catch { /* pipe closed — clearInterval below */ }
      }
    }, keepaliveMs);
  };

  try {
    for await (const item of iterable) {
      if (!headSent) {
        const detected = looksLikeHead(item);
        if (detected) {
          sendHead(detected.status, detected.headers);
          startHeartbeat();
          if (detected.body != null && detected.body !== '') {
            streamChunk(detected.body);
            lastEmit = Date.now();
          }
          continue;
        }
        sendHead(200, { 'Content-Type': 'text/plain; charset=utf-8' });
        startHeartbeat();
      }
      streamChunk(item);
      lastEmit = Date.now();
    }
    if (!headSent) sendHead(200, { 'Content-Type': 'text/plain; charset=utf-8' });
    writeFrame({ type: 'response_end' });
  } catch (err) {
    // EPIPE = client disconnected mid-stream. Stop iterating; the
    // worker continues serving subsequent requests if its stdin is
    // still open. Anything else we re-raise so the outer try/catch
    // emits an error frame BEFORE the head went out.
    if (err && err.code !== 'EPIPE') {
      if (!headSent) throw err;
      // Head already flew — nothing useful to surface. Best-effort end.
      try { writeFrame({ type: 'response_end' }); } catch {}
    }
  } finally {
    if (hbTimer) clearInterval(hbTimer);
  }
}

// drainReadableStream yields chunks from a fetch-API ReadableStream
// (e.g. `new Response(stream).body`). Used when the handler returns a
// Response whose body is a stream — we surface it as if the user had
// written an async generator yielding bytes.
async function* drainReadableStream(readable) {
  const reader = readable.getReader();
  try {
    for (;;) {
      const { done, value } = await reader.read();
      if (done) return;
      if (value) yield value;
    }
  } finally {
    try { reader.releaseLock(); } catch {}
  }
}

// readExactly reads exactly n bytes from process.stdin, resolving with null
// on EOF. Works by concatenating the available readable chunks.
function readExactly(n) {
  return new Promise((resolve, reject) => {
    const out = [];
    let have = 0;
    const stdin = process.stdin;

    const onReadable = () => {
      let chunk;
      while ((chunk = stdin.read(Math.min(n - have, 65536))) !== null) {
        out.push(chunk);
        have += chunk.length;
        if (have >= n) {
          cleanup();
          return resolve(Buffer.concat(out, have));
        }
      }
    };
    const onEnd = () => { cleanup(); resolve(null); };
    const onError = (err) => { cleanup(); reject(err); };
    const cleanup = () => {
      stdin.removeListener('readable', onReadable);
      stdin.removeListener('end', onEnd);
      stdin.removeListener('error', onError);
    };

    stdin.on('readable', onReadable);
    stdin.on('end', onEnd);
    stdin.on('error', onError);
    // In case data is already buffered.
    onReadable();
  });
}

async function readFrame() {
  const header = await readExactly(4);
  if (!header) return null;
  const len = header.readUInt32BE(0);
  if (len === 0) return {};
  const payload = await readExactly(len);
  if (!payload) return null;
  try {
    return JSON.parse(payload.toString('utf-8'));
  } catch (err) {
    return { type: 'request', event: { method: 'POST', path: '/', headers: {}, body: '' } };
  }
}

(async () => {
  // Optional recycle cap: after MAX_REQUESTS dispatches, exit so the pool
  // respawns and we avoid any slow memory creep in user code.
  const maxReqs = Number(process.env.ORVA_MAX_REQUESTS || 0);
  let served = 0;

  for (;;) {
    const frame = await readFrame();
    if (!frame) process.exit(0); // stdin EOF = clean shutdown
    if (frame.type === 'quit') {
      writeFrame({ type: 'bye' });
      process.exit(0);
    }
    if (frame.type !== 'request') continue;

    const event = frame.event || { method: 'POST', path: '/', headers: {}, body: '' };
    // Propagate call depth into env so orva.invoke()'s SDK can forward
    // it on outbound nested calls. Otherwise the host's depth guard
    // never trips on recursion.
    const _hdrs = event.headers || {};
    const _depth = _hdrs['x-orva-call-depth'] || _hdrs['X-Orva-Call-Depth'] || '';
    if (_depth) process.env.ORVA_CALL_DEPTH = _depth;
    else delete process.env.ORVA_CALL_DEPTH;

    // v0.5 trace context. Each event carries the trace_id + span_id of
    // this invocation; the SDK reads them from env when issuing nested
    // F2F calls or job enqueues so causal chains stay linked.
    const _tID = _hdrs['x-orva-trace-id'] || _hdrs['X-Orva-Trace-Id'] || '';
    const _sID = _hdrs['x-orva-span-id']  || _hdrs['X-Orva-Span-Id']  || '';
    if (_tID) process.env.ORVA_TRACE_ID = _tID; else delete process.env.ORVA_TRACE_ID;
    if (_sID) process.env.ORVA_SPAN_ID  = _sID; else delete process.env.ORVA_SPAN_ID;

    // v0.4 C1: streaming flag + heartbeat interval ride on per-request
    // headers so the proxy can flip them at runtime without redeploying
    // the worker. Defaults match the system_config seed values.
    const streamingOn = (_hdrs['x-orva-streaming-enabled'] ?? '1') !== '0';
    const keepaliveS = Math.max(1, Number(_hdrs['x-orva-stream-keepalive-seconds'] ?? '15') || 15);
    const keepaliveMs = keepaliveS * 1000;

    try {
      const result = await dispatch(event);

      // Streaming detection — async iterables, sync iterables, or a
      // Response whose body is a ReadableStream. Order matters:
      // Response objects are also iterable (ReadableStream is async-
      // iterable in Node 18+) so we MUST check the Response branch
      // first to extract status + headers, then drain the body stream.
      if (
        result &&
        typeof result === 'object' &&
        typeof result.status === 'number' &&
        result.body && typeof result.body.getReader === 'function'
      ) {
        const headers = {};
        if (result.headers && typeof result.headers.forEach === 'function') {
          result.headers.forEach((v, k) => { headers[k] = v; });
        }
        // Wrap in a generator that prepends the head, then yields chunks.
        async function* withHead() {
          yield { statusCode: result.status, headers, body: '' };
          for await (const chunk of drainReadableStream(result.body)) yield chunk;
        }
        await streamIterable(withHead(), streamingOn, keepaliveMs);
      } else if (
        result != null &&
        typeof result === 'object' &&
        (typeof result[Symbol.asyncIterator] === 'function' ||
         (typeof result[Symbol.iterator] === 'function' && typeof result !== 'string'))
      ) {
        // Exclude strings, Buffers, and arrays from being treated as
        // streaming iterables — those are perfectly valid response
        // bodies and shouldn't surprise the user.
        const isExcluded =
          typeof result === 'string' ||
          Buffer.isBuffer(result) ||
          Array.isArray(result) ||
          result instanceof Uint8Array;
        if (isExcluded) {
          const out = await normaliseResponse(result);
          writeFrame({ type: 'response', statusCode: out.statusCode, headers: out.headers, body: out.body });
        } else {
          await streamIterable(result, streamingOn, keepaliveMs);
        }
      } else {
        const out = await normaliseResponse(result);
        writeFrame({ type: 'response', statusCode: out.statusCode, headers: out.headers, body: out.body });
      }
    } catch (err) {
      process.stderr.write(`Handler error: ${err.stack || err.message}\n`);
      // If we already started streaming the response head, the proxy is
      // mid-loop and writeFrame on a fresh "response" envelope would
      // confuse it. Best-effort: just close. The proxy will surface the
      // truncation via the connection close.
      try {
        writeFrame({
          type: 'response',
          statusCode: 500,
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ error: 'Internal function error', message: err.message }),
        });
      } catch {}
    }

    served++;
    if (maxReqs > 0 && served >= maxReqs) {
      writeFrame({ type: 'bye' });
      process.exit(0);
    }
  }
})().catch((err) => {
  try {
    writeFrame({ type: 'error', fatal: true, message: String(err && err.stack || err) });
  } catch {}
  process.exit(1);
});
